### PR TITLE
feat: add account-level storage values

### PR DESCRIPTION
**Pull request by amatria**
### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed
Add account-level storage values:
```lua
Game.setAccountStorageValue(accountId, key, value)
Game.getAccountStorageValue(accountId, key)
-- force a save
Game.saveAccountStorageValues()
```

**Issues addressed:** 
 #3486

[cont]: https://github.com/otland/forgottenserver/wiki/Contributing
[code]: https://github.com/otland/forgottenserver/wiki/TFS-Coding-Style-Guide

### CODE REVIEW

**Comment by MillhioreBT** [comment]
It doesn't sound like a bad idea, I really like it a lot, I'm going to think about it a bit, and check your code ;)

**Comment by omarcopires** [comment]
What happens if I'm logged in with two users on the same account?

**Comment by MillhioreBT** [comment]
You'd better load the storages into an AccountStorageMap cache.

* We load the storages only if the player who enters is the first character of the account who enters.
* We save the storages only if the player who disconnects is the only character on the online account.

File: schema.sql [hunk 1]
```diff
@@ -96,6 +96,14 @@ CREATE TABLE IF NOT EXISTS `account_ban_history` (
   FOREIGN KEY (`banned_by`) REFERENCES `players` (`id`) ON DELETE CASCADE ON UPDATE CASCADE
 ) ENGINE=InnoDB DEFAULT CHARACTER SET=utf8;
 
+CREATE TABLE IF NOT EXISTS `account_storages` (
```
Comment (marmichalski) [inline schema.sql:99]: Please update db_version on line 360.

File: src/iologindata.cpp [hunk 2]
```diff
@@ -44,6 +44,32 @@ Account IOLoginData::loadAccount(uint32_t accno)
...
+	if (value != -1) {
+		query = fmt::format("INSERT INTO `account_storages` (`account_id`, `key`, `value`) VALUES ({:d}, {:d}, {:d}) ON DUPLICATE KEY UPDATE `value` = {:d}", accno, key, value, value);
+	} else {
+		query = fmt::format("DELETE FROM `account_storages` WHERE `account_id` = {:d} AND `key` = {:d}", accno, key);
+	}
+
+	return Database::getInstance().executeQuery(query);
+}
+
+bool IOLoginData::getAccountStorageValue(const uint32_t accno, const uint32_t key, int32_t& value)
```
Comment (marmichalski) [inline src/iologindata.cpp:60]: I don't think that returning the value by referenced parameter is good. Return value does not have to be a boolean at all, you can distinguish missing storage (-1) from set storage (>= 0) easily.

Anyway, this is probably true for players storage too, I don't see how mixing int32 with uint32 for value is acceptable.

**Comment by Zbizu** [comment]
Can't review from a phone, but basically:

account_storage (not storages) as database table for consistency with player_storage

File: src/iologindata.cpp (hunk 2 above)
Comment (amatria) [reply src/iologindata.cpp:60]: I agree with you indeed. Do you think it is a good idea I change the `player:setStorageValue` implementation too within this PR?

Comment (marmichalski) [reply src/iologindata.cpp:60]: I'd consider changing return type a breaking change and I have no idea what the policy here is (if any). This should probably be done in [2.0](https://github.com/otland/forgottenserver/projects/3).

**Comment by amatria** [comment]
So, as @MillhioreBT said. The proposed solution would be as follows:
1. Create a `AccountStorageRepository` class. This class will be in charge of storing a `storageMap` for each account logged in into the game. Also, it will store information about how many players from the same account are actually logged in into the game.
2. Once a player logs in into the game, it will push into the repository its account storage values.
3. If another player from the very same account logs in too, the repository will give him a reference to the already existing `storageMap`.
4. Players will now be able to consistently modify the account storages values.
5. Once all players from one account have logged out of the game, it will remove the shared reference from the repository and proceed to store it into the database.
Thoughts? Does it even make any sense to you? :p

**Comment by Zbizu** [comment]
it could be like this:

> map: accountId to account struct/object
> 
> account struct/object:
> - id
> - playersActive
> - storages(map: key to value)
> 
> login: map[accountId] exists and playersActive +1 or create map[accountId]
> logout: playersActive -1, save to database, if playersActive 0 then destroy

though I'm not sure how stable will that be when someone decides to spam relogs
on the other hand keeping the table in memory for entire server uptime might be ram consuming so I suggest adding lastUpdated field that would be helpful to determine whether the object can still be used soon or could be destroyed safely (and not destroying on logout).

**Comment by MillhioreBT** [comment]
I think it is better to have a map in the *Game class* and use this map to store the storages

<details>
<summary>iologindata.cpp</summary>

* load
```c++
//load account storage map
if (g_game.getPlayersByAccount(accountId).empty()) {
	if (result = db.storeQuery(fmt::format("SELECT `key`, `value` FROM `account_storage` WHERE `account_id` = {:d}", accountId))) {
		do {
			g_game.setAccountStorageValue(accountId, result->getNumber("key"), result->getNumber("value"));
		} while (result->next());
	}
}
```

* save
```c++
//save account storage map
if (g_game.getPlayersByAccount(accountId).size() == 1) {
	if (!db.executeQuery(fmt::format("DELETE FROM `account_storage` WHERE `account_id` = {:d}", accountId))) {
		return false;
	}

	const auto& accountStorageMap = g_game.accountStorageMap.find(accountId);
	if (accountStorageMap == g_game.accountStorageMap.end()) {
		return false;
	}

	query.str(std::string());
	DBInsert accountStorageQuery("INSERT INTO `account_storage` (`account_id`, `key`, `value`) VALUES ");
	for (const auto& it : accountStorageMap->second) {
		query << accountId << ',' << it.first << ',' << it.second;
		if (!accountStorageQuery.addRow(query)) {
			return false;
		}
	}

	if (!accountStorageQuery.execute()) {
		return false;
	}

	g_game.accountStorageMap.erase(accountId);
}
```
</details>

<details>
<summary>game.cpp</summary>

```c++
void Game::setAccountStorageValue(uint32_t accountId, const std::string key, const std::string& value)
{
	accountStorageMap[accountId][key] = value;
}

bool Game::getAccountStorageValue(uint32_t accountId, const std::string key, std::string& value) const
{
	auto accountMapIt = accountStorageMap.find(accountId);
	if (accountMapIt != accountStorageMap.end()) {
		auto storageMapIt = accountMapIt->second.find(key);
		if (storageMapIt == accountMapIt->second.end()) {
			value = "-1";
			return false;
		}

		value = storageMapIt->second;
		return true;
	}

	return false;
}
```
</details>

<details>
<summary>game.h</summary>

```h
// public:
std::unordered_map<uint32_t, StorageMap> accountStorageMap;
```

</details>

**Comment by amatria** [comment]
@MillhioreBT
And therefore change the Lua script interface methods to
```lua
Game.setAccountStorageValue(accId, key, value)
Game.getAccountStorageValue(accId, key)
```
right?

**Comment by amatria** [comment]
I have just squashed all the WIP commits and the PR should now be ready for a comprehensive review.

File: src/game.cpp [hunk 3]
```diff
@@ -4407,6 +4407,45 @@ void Game::addDistanceEffect(const SpectatorVec& spectators, const Position& fro
...
+		return;
+	}
+
+	accountStorageMap[accountId][key] = value;
+}
+
+bool Game::getAccountStorageValue(const uint32_t accountId, const uint32_t key, int32_t& value) const
+{
+	const auto& accountMapIt = accountStorageMap.find(accountId);
+
```
Comment (nekiro) [inline src/game.cpp:4423]: remove new line

File: src/game.cpp [hunk 4]
```diff
@@ -4407,6 +4407,45 @@ void Game::addDistanceEffect(const SpectatorVec& spectators, const Position& fro
...
+	}
+
+	value = -1;
+	return false;
+}
+
+size_t Game::getNumberOfPlayersByAccount(const uint32_t accountId) const
+{
+	size_t ret = 0;
+
```
Comment (nekiro) [inline src/game.cpp:4439]: remove new line

File: src/iologindata.cpp [hunk 5]
```diff
@@ -536,6 +536,15 @@ bool IOLoginData::loadPlayer(Player* player, DBResult_ptr result)
 		} while (result->next());
 	}
 
+	// load account storage map
+	if (g_game.getNumberOfPlayersByAccount(player->getAccount()) == 0) {
```
Comment (nekiro) [inline src/iologindata.cpp:540]: wouldn't find by account id be enough here? Why checking the players count at all?

File: src/iologindata.cpp [hunk 6]
```diff
@@ -832,6 +841,31 @@ bool IOLoginData::savePlayer(Player* player)
 		return false;
 	}
 
+	// save account storage values
+	if (g_game.getNumberOfPlayersByAccount(player->getAccount()) == 1) {
```
Comment (nekiro) [inline src/iologindata.cpp:845]: same here, just check if account id key exist and save it

**Review by nekiro — CHANGES_REQUESTED** [review]

File: src/iologindata.cpp (hunk 5 above)
Comment (amatria) [reply src/iologindata.cpp:540]: Indeed, it works when you load the account storage map for the first time. You call the `g_game.getPlayerByAccount(player->getAccount())` and you get a `nullptr` value. However, to be able to save the account storages only once, you need to know how many players are online from the very same account. In this case, the `g_game.getPlayerByAccount(player->getAccount())` call won't help you at all. You will just receive a reference to a player (which, in fact, can be the already logged out player). But, are there more online?

Comment (amatria) [reply src/iologindata.cpp:540]: Also, `g_game.getPlayerByAccount(player->getAccount())` is an O(n) function too [1]. So, I don't see any big improvement over using the new function either.

===
[1] Worst case scenario.

File: src/iologindata.cpp (hunk 6 above)
Comment (MillhioreBT) [reply src/iologindata.cpp:845]: What we want is to avoid saving the storage map several times, since we can save it only once, when in fact all the players of the same account have disconnected, otherwise every time a player disconnects we will be saving the storages, I suppose this is implemented like this in case several characters of the same account can connect at the same time

**Comment by MillhioreBT** [comment]
The idea is very good, and thanks for following my suggestions <3 It only remains to wait for the suggestions of the other developers, to know if they have better ideas, or perhaps a way to relocate the functions or their names, but in my opinion this is functional, it only remains to keep testing and wait!

File: schema.sql [hunk 7]
```diff
@@ -96,6 +96,14 @@ CREATE TABLE IF NOT EXISTS `account_ban_history` (
   FOREIGN KEY (`banned_by`) REFERENCES `players` (`id`) ON DELETE CASCADE ON UPDATE CASCADE
 ) ENGINE=InnoDB DEFAULT CHARACTER SET=utf8;
 
+CREATE TABLE IF NOT EXISTS `account_storage` (
+	`account_id` int NOT NULL,
```
Comment (EPuncker) [inline schema.sql:100]: indentation in sql file is made using spaces

**Comment by marmichalski** [comment]
I do not believe that those "micro-optimizations" are of any good. There are a lot of other places that executes bigger sql queries instantly. MySQL is not a bottleneck here, those are simple insert and select statements and current solution introduces synchronization problems when you'd like to use account_storage in, for example, your web app. Saving account_storage when there is only one player online per-account is IMHO also wrong, consider the web app sync + server allowing MC on the same account scenario.

tl;dr; no-cache should be the way to go IMHO, account storage was implemented in many servers as instant queries and generated no problems, one of those was Classicus with hundreds of players.

**Comment by amatria** [comment]
>I do not believe that those "micro-optimizations" are of any good

I would not consider them "micro-optimizations". Indeed, I would say that is a rather bold claim. I know MySQL implements caching mechanisms to optimize fetching and pulling data in and out of the relations. However, is it even as fast as accessing a hash table in user space? Also, what if the game server was sitting in a DMZ and the database was running behind a firewall to prevent attacks? Would you still consider this solution a "micro-optimization". It may sound bizarre, but it is a completely feasible scenario.

Nevertheless, I get your point. I liked the queries approach too. It was dead simple, clean and effective. So... I guess we have to see what the other developers think about the matter.

>current solution introduces synchronization problems when you'd like to use account_storage in, for example, your web app

And so it does for player storages. Doesn't it?

**Comment by marmichalski** [comment]
Sure it's not that fast, but does it have to be? It's fast enough. 

Check iomarket. Yes, I'd still consider this micro optimizations compared to other "instant" queries. I'm ignoring the DMZ part, as probably many people will have to google what that even means 😉

Player storages are at least saved with each player save.

**Comment by amatria** [comment]
>Sure it's not that fast, but does it have to be? It's fast enough.

Gotchu :p. To be honest, I'm with you with the queries approach (and there's no sync problem), even though I like the other solution a lot too. So, well, let's see what the others have to say. 

>Player storages are at least saved with each player save.

Yeah, they are. But they also follow the same DELETE and INSERT pattern as in this solution. Therefore, if a value is stored from outside of the C++ code, it will overwrite/get rid of the result stored directly into the database.

**Comment by Zbizu** [comment]
A possible sync system could be an agreement between aac developers and us.

The aac can have setAccountStorage method in PHP, a method that would queue the storages for updating and the server could pick, let's say, 20 results every 15 seconds. This solution is already used for website shop systems.

**Comment by marmichalski** [comment]
HI @amatria, I'm back after a bad sleep 😎
I did not mean to sound hostile, so if I did, I apologise.

> 
> > Player storages are at least saved with each player save.
> 
> Yeah, they are. But they also follow the same DELETE and INSERT pattern as in this solution. Therefore, if a value is stored from outside of the C++ code, it will overwrite/get rid of the result stored directly into the database.

That's a side-effect I did not even think about. Imagine web-shop account-wide offer, think outfits. Account storage could be leveraged for that.  
But that's now what my original thought of "desync" was about. There are servers that never restart (Cyntara?), allowing multiple characters to be online on the same account (Cyntara?) and AFK-botting/training/whatever (Cyntara?). Now, let's have consider a situation, where somebody is "abusing" these things and is online for hundreds of hours - his account storage never persists into db.

I know, this might be super-rare situation, so let me introduce another one: saving server in gdb after crash.
Sometimes (always?) you want to call g_game.saveGameState after a crash and (most of the times?) it works. Now, there is more than 1 character online per many accounts, none of those accounts storage will persist.

So, consider the following changes:
* Loading storage:
  * Always try to load account storage, if the entry is already set, no-op
* Saving storage:
  * Perhaps introduce "dirty" state for AccountStorage and make every change to that (new?) object marking it as dirty
  * Save when dirty, reset dirty flag, therefore it will be saved only once, even if there are 100 characters online from that account
* [Optionally] If memory is so precious, cleanup the AccountStorage object on-logout (or player destructor?)

What does it solve? Iterating all online players when loading/saving one. Preventing storage being not loaded/saved in certain scenarios.

- Also, you are erasing the storages on save - but save does not mean that player logs out 🤔 
- "Virtual" players are not supported
- dunno, did not review whole code 🍨 

Have a nice day.

File: src/game.cpp [hunk 8]
```diff
@@ -4407,6 +4407,43 @@ void Game::addDistanceEffect(const SpectatorVec& spectators, const Position& fro
...
+size_t Game::getNumberOfPlayersByAccount(const uint32_t accountId) const
+{
+	size_t ret = 0;
+	for (const auto& it : players) {
+		if (it.second->getAccount() == accountId) {
+			ret++;
+		}
+	}
+
+	return ret;
```
Comment (ranisalt) [inline src/game.cpp:4444]: Idea:

```
	return std::count_if(players.begin(), players.end(), [=](auto& player) { return player->getAccount() == accountId });
```

File: src/game.cpp [hunk 9]
```diff
@@ -4407,6 +4407,43 @@ void Game::addDistanceEffect(const SpectatorVec& spectators, const Position& fro
...
+{
+	if (value == -1) {
+		accountStorageMap[accountId].erase(key);
+		return;
+	}
+
+	accountStorageMap[accountId][key] = value;
+}
+
+bool Game::getAccountStorageValue(const uint32_t accountId, const uint32_t key, int32_t& value) const
```
Comment (ranisalt) [inline src/game.cpp:4420]: Return a `pair<bool, int32_t>` or just return `-1` as it was defined to be the invalid value

**Review by ranisalt — CHANGES_REQUESTED** [review]

File: src/iologindata.cpp (hunk 6 above)
Comment (ranisalt) [reply src/iologindata.cpp:845]: > otherwise every time a player disconnects we will be saving the storages

Is this a problem? If the server crashes, at least *some* of the values will be updated from the session. It's a write-through cache of sorts...

File: src/iologindata.cpp (hunk 5 above)
Comment (ranisalt) [reply src/iologindata.cpp:540]: >  [1] Worst case scenario.

It is actually always linear, but the problem persists. Is there nothing on a player object that tags it is online or not?

File: src/game.cpp (hunk 8 above)
Comment (MillhioreBT) [reply src/game.cpp:4444]: yeah, better code ^ thanks Rani you always beautiful :p

File: src/iologindata.cpp (hunk 6 above)
Comment (amatria) [reply src/iologindata.cpp:845]: Okay. I have to agree with you. We can get rid of this function and just go like:
1. Load only once using `Game::getPlayerByAccount(uint32_t accountId)` [1].
2. Always store the values.

===
[1] Checking if a player is online is of no use.

Comment (amatria) [reply src/iologindata.cpp:845]: Okay. I like the write-through idea. However, we can't get rid of this function just yet. We need it in order to delete the account storage map object, once all players have logged out of the account.

To sum up, I believe the final solution would be as follows (I'll update the branch right away):
1. Load the values either using the new function or `Game::getPlayerByAccount(uint32_t acc) == nullptr`. And actually, I'll go with the second option. It is not exactly an O(n) operation when there are one or more players from the same account logged in into the game.
2. Always save the storages.
3. Delete the account storage map reference using the `Game::getNumPlayersByAccount(uint32_t acc)` method.

**Comment by amatria** [comment]
@marmichalski 
>I did not mean to sound hostile, so if I did, I apologise.

No, you didn't. No worries :p.

As for the topic. I believe that implementing a dirty state is far too complex for this use case. Also, with the latest commit this problem is long gone. We simply save the values every time a player logs out. And we needn't to implement a dirty cache. I appreciate your extensive response though. Time is valuable. Not that many people would write such a comprehensive and educated comment.

Take very good care of yourself ❤️.

File: src/iologindata.cpp [hunk 10]
```diff
@@ -832,6 +841,31 @@ bool IOLoginData::savePlayer(Player* player)
...
+			return false;
+		}
+	}
+
+	if (!accountStorageQuery.execute()) {
+		return false;
+	}
+
+	if (g_game.getNumberOfPlayersByAccount(player->getAccount()) == 1) {
+		g_game.accountStorageMap.erase(player->getAccount());
```
Comment (marmichalski) [inline src/iologindata.cpp:866]: This is not a place to do it, as I have mentioned in my other comment. Calling this method does not mean that player is logging out. You are erasing the storage with every save of that player (player:save(), saveServer()).

Comment (amatria) [reply src/iologindata.cpp:866]: True. I missed that one (twice, lmao).

**Comment by nekiro** [comment]
Let's stop microoptimizing and just make storages save in game class, load it all once on startup and then just save it all once on shutdown. No problems with keepin track of anything and it works just like that.

**Comment by MillhioreBT** [comment]
I was thinking that this should not be implemented, as it can be done perfectly through pure Lua!
However, @nekiro idea is good, although I would like them to be global storages and there we would store the account storages
The current global storage methods do not make sense XD, nobody uses them because they are useless `Game.setStorageValue / Game.getStorageValue`, these methods should be in the sources and have a real storage in the database as before, in this way you can do something like what I did on my server:

```lua
function Player:setAccountStorageValue(key, value)
	return Game.setStorageValue(("AccountStorage%s:%s"):format(self:getAccountId(), key), value)
end

function Player:getAccountStorageValue(key)
	return Game.getStorageValue(("AccountStorage%s:%s"):format(self:getAccountId(), key))
end
```

**Comment by Zbizu** [comment]
@MillhioreBT tbh the longer this discussion goes on, the more I agree with you regarding writing that purely in lua

**Comment by amatria** [comment]
Well, I just pushed the changes nekiro proposed. However, if you want to change the whole implementation to pure Lua, I'd rather close this PR and open a new one. So... It's up to you. I like how this one turned out eventually.

**Comment by MillhioreBT** [comment]
> Well, I just pushed the changes nekiro proposed. However, if you want to change the whole implementation to pure Lua, I'd rather close this PR and open a new one. So... It's up to you. I like how this one turned out eventually.

Let someone else review this, and it depends on what the PRO's say, you make the decision to close the PR and do something simpler in lua, or just do something better, and I apologize for writing so much and extending this discussion so much

**Comment by amatria** [comment]
What's good people 🌚! It's been 10 days since I last updated this PR. I understand it's summer and we all are busy doing RL stuff etc. However, I would appreciate if further action was taken. Are we closing this PR and implementing the changes in Lua or approving it instead?

Have a lovely weekend :p.

**Review by DSpeichert — APPROVED** [review]

**Review by nekiro — APPROVED** [review]
LGTM

File: src/game.cpp [hunk 11]
```diff
@@ -4407,6 +4412,73 @@ void Game::addDistanceEffect(const SpectatorVec& spectators, const Position& fro
...
+		return false;
+	}
+
+	if (!db.executeQuery("DELETE FROM `account_storage`")) {
+		return false;
+	}
+
+	for (const auto& accountIt : g_game.accountStorageMap) {
+		if (accountIt.second.empty()) {
+			break;
```
Comment (MillhioreBT) [inline src/game.cpp:4464]: If you break the loop here, then the other accounts will not be saved, you should use continue
